src/main/cpp/src/cast_string_to_datetime.cu

```diff
@@ -858,6 +861,163 @@ std::unique_ptr<cudf::column> parse_ts_strings(cudf::strings_column_view const&
+  bool negative_year_sign = false;
+  char const sign_c       = ptr[pos];
+  if ('-' == sign_c || '+' == sign_c) {
+    ++pos;
+    if ('-' == sign_c) { negative_year_sign = true; }
+  }
+
+  // parse year: yyyy[y][y][y]
+  if (!parse_int(ptr,
+                 pos,
+                 end_pos,
+                 d.year,
+                 /*min_digits*/ 4,
+                 /*max_digits*/ 7)) {
+    return false;
+  }
+  if (negative_year_sign) { d.year = -d.year; }
+
+  if (eof(pos, end_pos)) {
+    // only has: yyyy[y][y][y], return early
+    return true;
+  }
+
+  // parse month: -[m]m
+  if (!parse_char(ptr, pos, '-') || !parse_int(ptr,
+                                               pos,
+                                               end_pos,
+                                               d.month,
+                                               /*min_digits*/ 1,
+                                               /*max_digits*/ 2)) {
+    return false;
+  }
+
+  if (eof(pos, end_pos)) {
+    // only has: yyyy[y][y]-[m]m, return early
+    return true;
+  }
+
+  // parse day: -[d]d
+  if (!parse_char(ptr, pos, '-') || !parse_int(ptr,
+                                               pos,
+                                               end_pos,
+                                               d.day,
+                                               /*min_digits*/ 1,
+                                               /*max_digits*/ 2)) {
+    return false;
+  }
+
+  if (eof(pos, end_pos)) {
+    // no tailing characters, e.g.: "2025-01-01"
+    return true;
+  }
+
+  // parse date time separator
+  // ignore the tailing chars after ' ' or 'T', e.g.: the following are valid:
+  // "2025-01-01T", "2025-01-01Txxxx", and "2025-01-01 xxx"
+  return parse_date_time_separator(ptr, pos);
+}
+
+struct parse_string_to_date_fn {
+  // input strings
+  cudf::column_device_view d_strings;
+
+  // output columns
+  bool* validity;
+  cudf::timestamp_D* output;
+
+  __device__ void operator()(cudf::size_type const idx) const
+  {
+    // check null
+    if (d_strings.is_null(idx)) {
+      validity[idx] = false;
+      return;
+    }
+
+    auto const str                   = d_strings.element<cudf::string_view>(idx);
+    unsigned char const* str_ptr     = reinterpret_cast<unsigned char const*>(str.data());
+    unsigned char const* str_end_ptr = str_ptr + str.size_bytes();
+
+    // parse the date string to segments
+    spark_rapids_jni::date_segments date_segments;
+    auto result_success = parse_date(str_ptr, str_end_ptr, date_segments);
+
+    // check parsed result
+    if (!result_success || !date_segments.is_valid_date()) {
+      validity[idx] = false;
+      return;
+    }
+
+    // calculate the epoch day and check 'days' fits in int32_t
+    int64_t days  = date_segments.to_epoch_day();
+    output[idx]   = cudf::timestamp_D{cudf::duration_D{static_cast<int32_t>(days)}};
+    validity[idx] = days >= cuda::std::numeric_limits<int32_t>::min() &&
```

Comment:
With the int32 limitation, the year > `Int.MaxValue / 365 = 5883516` will return null ?

Reply:
Yes. Exactly year > (Int.MaxValue / 365 - 1970), 1970-01-01 is the epoch day.

Reply:
Looks good to me if Spark also returns null